src/Storages/ObjectStorage/StorageObjectStorageSource.cpp

```diff
@@ -62,6 +63,98 @@ namespace ErrorCodes
     extern const int FILE_DOESNT_EXIST;
 }
 
+struct CachedGlobObjectIterator : IObjectIterator, WithContext
+{
+    using Configuration = StorageObjectStorage::Configuration;
+    using ConfigurationPtr = StorageObjectStorage::ConfigurationPtr;
+    using ObjectInfos = StorageObjectStorage::ObjectInfos;
+
+    CachedGlobObjectIterator(
+        const std::shared_ptr<ObjectInfos> & object_infos_,
+        ConfigurationPtr configuration_,
+        const ActionsDAG::Node * predicate,
+        const NamesAndTypesList & virtual_columns,
+        ContextPtr context_,
+        std::size_t ,
+        ObjectInfos * read_keys = nullptr)
+        : WithContext(context_),
+        object_infos(object_infos_),
+        configuration(configuration_)
+    {
+        if (configuration->isPathWithGlobs())
+        {
+            const auto key_with_globs = configuration->getPath();
+
+            //
+            auto matcher = std::make_unique<re2::RE2>(makeRegexpPatternFromGlobs(key_with_globs));
+            if (!matcher->ok())
+            {
+                throw Exception(ErrorCodes::CANNOT_COMPILE_REGEXP, "Cannot compile regex from glob ({}): {}", key_with_globs, matcher->error());
+            }
+
+            ExpressionActionsPtr filter_expr;
+            recursive = key_with_globs == "/**";
+            if (auto filter_dag = VirtualColumnUtils::createPathAndFileFilterDAG(predicate, virtual_columns))
+            {
+                VirtualColumnUtils::buildSetsForDAG(*filter_dag, getContext());
+                filter_expr = std::make_shared<ExpressionActions>(std::move(*filter_dag));
+            }
+
+            for (auto it = object_infos->begin(); it != object_infos->end();)
+            {
+                if (!recursive && !re2::RE2::FullMatch((*it)->getPath(), *matcher))
+                    it = object_infos->erase(it);
+                else
+                    ++it;
+            }
+
+            if (filter_expr)
+            {
+                std::vector<String> paths;
+                paths.reserve(object_infos->size());
+                for (const auto & object_info : *object_infos)
+                    paths.push_back(StorageObjectStorageSource::getUniqueStoragePathIdentifier(*configuration, *object_info, false));
+
+                VirtualColumnUtils::filterByPathOrFile(*object_infos, paths, filter_expr, virtual_columns, getContext());
+
+                LOG_TEST(getLogger("Arthur"), "Filtered files: {} -> {}", paths.size(), object_infos->size());
+            }
+
+            if (read_keys)
+                read_keys->insert(read_keys->end(), object_infos->begin(), object_infos->end());
+        }
+        else
+        {
+            throw Exception(
+                ErrorCodes::BAD_ARGUMENTS,
+                "Using glob iterator with path without globs is not allowed (used path: {})",
```

Comment:
Shall not it be LOGICAL_ERROR ?
This looks like a branch of code where we cannot get normally (user does not select which iterator to use manually)

Reply:
I agree, it should probably be `LOGICAL_ERROR`. But:

This is mostly a copy and paste from the existing [GlobIterator](https://github.com/ClickHouse/ClickHouse/blob/caa99b4237f69f2d18b0b2416f4196c5cbad5150/src/Storages/ObjectStorage/StorageObjectStorageSource.cpp#L733).

I might refactor this to avoid duplication. For now, this is just a draft implementation.

Even if I refactor this, I would opt for keeping parity with existing code & upstream. This will make the review and merges with upsteam easier